File: archive/flutter-app/lib/core/utility/json_formatter.dart
import 'dart:convert';

import '../ports/i_tool_use_case.dart';

enum JsonFormatMode { pretty, minify, validate }

class JsonFormatterInput {
  const JsonFormatterInput({
    required this.source,
    required this.mode,
    this.indent = '  ',
  });

  final String source;
  final JsonFormatMode mode;
  final String indent;
}

class JsonFormatterResult {
  const JsonFormatterResult({required this.isValid, this.output, this.errorMessage});

  final bool isValid;
  final String? output;
  final String? errorMessage;
}

class JsonFormatter implements IToolUseCase<JsonFormatterInput, JsonFormatterResult> {
  const JsonFormatter();

  @override
  JsonFormatterResult execute(JsonFormatterInput input) {
    dynamic decoded;
    try {
      decoded = jsonDecode(input.source);
    } on FormatException catch (e) {
      return JsonFormatterResult(isValid: false, errorMessage: 'Invalid JSON: ${e.message}');
    }

    switch (input.mode) {
      case JsonFormatMode.validate:
        return const JsonFormatterResult(isValid: true, output: 'Valid JSON');
      case JsonFormatMode.pretty:
        final output = JsonEncoder.withIndent(input.indent).convert(decoded);
        return JsonFormatterResult(isValid: true, output: output);
      case JsonFormatMode.minify:
        final output = const JsonEncoder().convert(decoded);
        return JsonFormatterResult(isValid: true, output: output);
    }
  }
}

File: app/src/core/utility/jsonFormatter.ts
import type { IToolUseCase } from '../ports/IToolUseCase'

export type JsonFormatMode = 'pretty' | 'minify' | 'validate'

export interface JsonFormatterInput {
  source: string
  mode: JsonFormatMode
  indent?: string
}

export interface JsonFormatterResult {
  isValid: boolean
  output?: string
  errorMessage?: string
}

export class JsonFormatter implements IToolUseCase<JsonFormatterInput, JsonFormatterResult> {
  execute(input: JsonFormatterInput): JsonFormatterResult {
    let decoded: unknown
    try {
      decoded = JSON.parse(input.source)
    } catch (e) {
      const message = e instanceof Error ? e.message : String(e)
      return { isValid: false, errorMessage: `Invalid JSON: ${message}` }
    }

    switch (input.mode) {
      case 'validate':
        return { isValid: true, output: 'Valid JSON' }
      case 'pretty':
        return { isValid: true, output: JSON.stringify(decoded, null, input.indent ?? '  ') }
      case 'minify':
        return { isValid: true, output: JSON.stringify(decoded) }
    }
  }
}
